lib/options.go

```diff
@@ -139,29 +139,44 @@ type TLSAuthFields struct {
 
 // Defines a TLS client certificate to present to certain hosts.
 type TLSAuth struct {
-	TLSAuthFields
-	certificate *tls.Certificate
+	Origin            []*TLSAuthFields
+	Certificates      []tls.Certificate
+	NameToCertificate map[string]*tls.Certificate
 }
 
 func (c *TLSAuth) UnmarshalJSON(data []byte) error {
-	if err := StrictJSONUnmarshal(data, &c.TLSAuthFields); err != nil {
+	if err := StrictJSONUnmarshal(data, &c.Origin); err != nil {
 		return err
 	}
-	if _, err := c.Certificate(); err != nil {
+	if err := c.FillData(); err != nil {
 		return err
 	}
 	return nil
 }
 
-func (c *TLSAuth) Certificate() (*tls.Certificate, error) {
-	if c.certificate == nil {
-		cert, err := tls.X509KeyPair([]byte(c.Cert), []byte(c.Key))
+func (c *TLSAuth) MarshalJSON() ([]byte, error) {
+	return json.Marshal(c.Origin)
+}
+
+func (c *TLSAuth) FillData() error {
+	c.Certificates = make([]tls.Certificate, len(c.Origin))
+	nameToCert := make(map[string]*tls.Certificate)
+	for i, auth := range c.Origin {
+		cert, err := tls.X509KeyPair([]byte(auth.Cert), []byte(auth.Key))
 		if err != nil {
-			return nil, err
+			return err
 		}
-		c.certificate = &cert
+		c.Certificates[i] = cert
+		for _, name := range auth.Domains {
+			nameToCert[name] = &cert
+		}
+	}
```

Comment:
We should move this directly into the UnmarshalJSON method. It isn't expected to use in other places.